Add --build-plan for 'cargo build'

mshal opened the pull request:
With 'cargo build --build-plan', cargo does not actually run any
commands, but instead prints out what it would have done in the form of
a JSON data structure.

Fixes #3815

aturon commented:
cc @aidanhs

matklad commented:
Awesome @mshal! I really like how this looks now!

I think we need to be really careful not to expose internal implementation details of Cargo, and this approach achieves this almost perfectly. I especially like how `BuildPlan` is a just a bunch of PODs. However, the following things leak some information:

* `buildkey` stores textual representation of various internal structures. It is intended to be used as opaque key, so this doesn't strictly matter, but perhaps there's a simple way to get more opaque id. For example, we could assign numerical ids to units, by lazily populating a `HashMap<Unit, usize>` map. 

* we expose `target` directory structure via various command-line options. I think this OK probably, as long as we document that precise outputs location are subject to change. I expect that, if a build systems wants to use a different directory structure, it can reliably post-process the build plan and substitute target dir with something else. Perhaps it makes sense to add a `targer_dir` field to build-plan itself, to make it easier to distinguish invariant source paths from potentially changeable target paths? 

* we expose `links` key. Is this really necessary? I think that we use it cargo mainly for user's convenience, and otherwise everything else works with unliked outputs from `deps` (not sure about this though, let's ask @alexcrichton).


Regarding the implementation itself, it looks great! We use slightly different code style in Cargo though, so you might want to `rustup +stable component add rustfmt-preview` and run `rustfmt +stavle --skip-children $FILE` for changed files. 

It would also be great to have some tests for this as well! I think we should create a `tests/testsuite/build_plan.rs` file and add a couple of smoke test. Here's an example of a test for JSON output: https://github.com/rust-lang/cargo/blob/b0d65af158bc2f2e4c1bc4e143a7dcd4b40af0f5/tests/testsuite/build.rs#L4024. Note how `[..]` and `{..}` are used as wildcards to match arbitrary strings/objects to get around platform specific variations.

bors commented:
:umbrella: The latest upstream changes (presumably #5348) made this pull request unmergeable. Please resolve the merge conflicts.

djc commented:
Sorry, I'm afraid my refactoring in #5348 has bitrotted this quite a bit. :(

What would be a nice side goal for this (though I hope that doesn't complicate it too much) is if cargo itself is refactored such that it goes through (a) generating the build plan, and then (b) executing from the build plan. That would make it more robust in the sense that we know the cargo build itself is not inadvertently relying on anything but the build plan.

That's not in this PR, as far as I can see, right?

mshal commented:
> buildkey stores textual representation of various internal structures. It is intended to be used as opaque key, so this doesn't strictly matter, but perhaps there's a simple way to get more opaque id. For example, we could assign numerical ids to units, by lazily populating a HashMap<Unit, usize> map.

I was trying to make the buildkey to be something human-readable, so for example you would be able to see that a "app_units v0.6.0 Target(lib) Profile(build) Target" depends on "num-traits v0.1.43 Target(lib) Profile(build) Target" and "serde v1.0.37 Target(lib) Profile(build) Target" by looking at the generated JSON output. This also gives the external build system some indication of what it is building, so it can use the key when displaying what it's building. Eg: "Now building app_units v0.6.0 Target(lib) Profile(build) Target".  Note this doesn't need to be consistent among cargo versions, as long as the buildkey is unique for different modules (including host vs. target vs. build-script iterations of a particular cargo module) during a single 'cargo --build-plan' invocation. I'm happy to use another naming scheme though if there's a better way to get this information.

> we expose target directory structure via various command-line options. I think this OK probably, as long as we document that precise outputs location are subject to change. I expect that, if a build systems wants to use a different directory structure, it can reliably post-process the build plan and substitute target dir with something else. Perhaps it makes sense to add a targer_dir field to build-plan itself, to make it easier to distinguish invariant source paths from potentially changeable target paths?

Sorry I don't quite follow this. What target are you referring to? Is this the --target flag in the json output? Or the self.target in the buildkey definition?

I don't think it matters if the output location changes, since the "outputs" field of the build plan explicitly lists where the file will be created. So if cargo decides things need to be in a different directory, a new 'cargo --build-plan' invocation will list a different output for a module, where the new file can be created.

> we expose links key. Is this really necessary? I think that we use it cargo mainly for user's convenience, and otherwise everything else works with unliked outputs from deps (not sure about this though, let's ask @alexcrichton).

I'm pretty sure this is necessary, otherwise the external build system won't know where to create the symlink for the "installed" version of the file.

> Regarding the implementation itself, it looks great! We use slightly different code style in Cargo though, so you might want to rustup +stable component add rustfmt-preview and run rustfmt +stavle --skip-children $FILE for changed files.

Ahh, thanks for the heads up! I've used this on the files I've changed, though I had to use '--toolchain stable' instead of '+stable'.

> It would also be great to have some tests for this as well! I think we should create a tests/testsuite/build_plan.rs file and add a couple of smoke test.

Good idea, I'll add some tests to get at least some basic coverage.

mshal commented:
> What would be a nice side goal for this (though I hope that doesn't complicate it too much) is if cargo itself is refactored such that it goes through (a) generating the build plan, and then (b) executing from the build plan. That would make it more robust in the sense that we know the cargo build itself is not inadvertently relying on anything but the build plan.
>
> That's not in this PR, as far as I can see, right?

I like that idea, but unfortunately that's a bit beyond my level of expertise at the moment. You are correct that it is not in this PR. Maybe we can file a followup issue? I'm not really sure what would be involved.

alexcrichton commented on src/cargo/ops/cargo_rustc/build_plan.rs:
```diff
@@ -0,0 +1,105 @@
+    program: String,
+    args: Vec<String>,
+    env: BTreeMap<String, String>,
+    cwd: String,
```
I think `PathBuf` implements `Serialize`, so perhaps that could be used here instead of `String` to avoid some unwraps and `to_str`?

mshal replied:
Looks like PathBuf works fine for cwd- thanks!

matklad commented:
>I was trying to make the buildkey to be something human-readable, so for example you would be able to see that a "app_units v0.6.0 Target(lib) Profile(build) Target" depends on "num-traits v0.1.43 Target(lib) Profile(build) Target" and "serde v1.0.37 Target(lib) Profile(build) Target" by looking at the generated JSON output. This also gives the external build system some indication of what it is building, so it can use the key when displaying what it's building.

Hm, interesting! I think it makes sense to also make it possible to connect build plan with other pieces of JSON information: `cargo metadata` and `cargo build --message-format=json`. To do that, we could include `PackageId` and `Target` field, [like we do for artifacts](https://github.com/rust-lang/cargo/blob/0a1add2d8689ad12a86f6c32d0a5cd0393dc5d80/src/cargo/util/machine_message.rs#L31-L32). Hm, perhaps we should just embed `Artifact` as a field of `Module`? 


>Yeah this is something we'll definitely want, otherwise Cargo I believe will only generate files with unreadable hashes!

>I'm pretty sure this is necessary, otherwise the external build system won't know where to create the symlink for the "installed" version of the file.

I have a slight suspicion that we are not on the same page here, so let me ask a clarifying question:

Am I correct that `links` is not strictly *required*? That is, I think if one executes the build plan without `link`s, everything just works because Cargo only ever uses stuff from the `deps` folder? That is, `link` is  there strictly for user's convenience, so that they can `./target/debug/my-binary`. 


The external build system would probably want to put final artifacts elsewhere, so the `link` key is probably not to relevant for it?

alexcrichton commented:
@matklad hm a good question. So I know a lot of artifacts are `target/debug/deps/foo` which get linked to `target/debug/foo`, but I think we've *also* got artifacts like `target/debug/deps/bar-xxxxxxx` which get linked to `target/debug/bar`. For those with hashes I think we need a `links`, right?

matklad commented:
>For those with hashes I think we need a links, right?

Hm, so I've made an experiment by disabling all linking, and looks like it is required at least for build scripts: 

```
error: failed to run custom build command for `regex v0.2.10`
could not execute process `/home/matklad/projects/cargo/target/debug/build/regex-c000d0f4c53b8f51/build-script-build` (never executed)
warning: build failed, waiting for other jobs to finish...
error: build failed
```

alexcrichton commented:
Heh interesting! We could probably fix Cargo to not actually require `links`, but I think we'd still want to keep it for the final artifact regardless

mshal commented:
> Would it be possible perhaps to do rules name foo-h145135135? The stringified version may not be entirely lossless in terms of capturing all the nuances that Cargo has, but hashing everything will for sure catch everything. I'm not sure if that's too unreadable of a target though? We could certainly work to make the prefix string pretty human readable.

In order to get the test cases working, I had to change the key to be something other than the buildkey, so I've tried to implement @matklad's HashMap<usize> suggestion. (The with_json() function doesn't seem to let me wildcard keys with things like "foo [..]", only values). I couldn't get it to work with Unit since we don't have a Unit available during the drain_the_queue() function where I call plan.update(), so I've left it as a buildkey String for now. If there's a good way to fix get_id() to use a Unit everywhere, I could use some suggestions. The previous key is available as "display_name" in the Module structure so it is more clear what it is used for.

> I see build plans as a sort of "build me a plan for this exact one invocation" rather than "build me a plan which is configurable at build-time in the same way Cargo is". For the latter case I think you'd want to generate multiple plans or just use Cargo.

Yes, exactly. If the cargo configuration changes (via Cargo.toml changes, environment variable changes, etc), we'd run cargo --build-plan to generate a new plan.

> Hm, interesting! I think it makes sense to also make it possible to connect build plan with other pieces of JSON information: cargo metadata and cargo build --message-format=json. To do that, we could include PackageId and Target field, like we do for artifacts. Hm, perhaps we should just embed Artifact as a field of Module?

Perhaps? It does look similar, though I don't have a full grasp of the overall architecture to know if they would easily be combined.

mshal commented:
The new push includes:
 - A few basic test cases for --build-plan
 - A module_map to use integer keys for the modules
 - Separate display_name for the user-readable string of the module
 - Ran rustfmt & rebased

mshal commented:
I forgot to ask:

1) Is there a way I can use serde's to_string_pretty() and still have it work with test cases? The with_json() function doesn't seem to like the fact that to_string_pretty() prints on multiple lines.

2) When usize is serialized, it gets serialized as a string when it's a key, but as an integer when it's in a Vec. For example, in BuildPlan I have modules: BTreeMap<usize, Module>, which ends up looking like this:

"0": { ... }

And inside the module, the deps: Vec<usize> show up like this:

    "deps": [2, 3, 4]

It would be helpful if they were either both quoted, or both not quoted, if possible.

matklad commented on src/cargo/core/compiler/build_plan.rs:
```diff
@@ -0,0 +1,136 @@
+
+    pub fn add_output(&mut self, path: &PathBuf, link: &Option<PathBuf>) {
+        self.outputs.push(path.clone());
+        if link.is_some() {
```
This could be `if let Some(ref link) = *link` to avoid `unwrap` down the line.

matklad commented on src/cargo/core/compiler/build_plan.rs:
```diff
@@ -0,0 +1,136 @@
+            .expect("unicode program string required")
+            .to_string()
+            .clone();
+        self.cwd = cmd.get_cwd().unwrap().to_path_buf();
```
I am not entirely sure we set `cwd` in call cases. Let's change `cwd: PathBuf,` to `cwd: Option<PathBuf>,`?

matklad commented on src/cargo/core/compiler/build_plan.rs:
```diff
@@ -0,0 +1,136 @@
+                var.clone(),
+                value
+                    .as_ref()
+                    .expect("environment value required")
```
I think this `expect` might panic sometimes. We use `None` to signify that the environment variable should be cleared, and by default we inherit env vars.

matklad commented on src/cargo/core/compiler/build_plan.rs:
```diff
@@ -0,0 +1,136 @@
+#[derive(Debug, Serialize)]
+pub struct BuildPlan {
+    module_map: BTreeMap<String, usize>,
+    modules: BTreeMap<usize, Module>,
```
I think we can make this `modules: Vec<Module>`, and then, in `get_id`, we could push a `Module::default` to this `Vec`. This will require `derive(Default)` as well.

matklad commented on src/cargo/core/compiler/context/mod.rs:
```diff
@@ -67,6 +69,27 @@ pub struct Unit<'a> {
+            "{} {} {} {:?}",
+            self.pkg, self.target, self.profile, self.kind
+        )
+    }
```
Let's switch to @alexcrichton suggestions with hashes? It could look like this I think:

```
impl<'a> Unit<'a> {
    pub fn buildkey(&self) -> String {
        #![allow(deprecated)] // for SipHash
        use std::hash::{Hash, Hasher, SipHasher};

        let mut hasher = SipHasher::new_with_keys(0, 0);
        self.hash(&mut hasher);
        format!("{}-{}", self.pkg.name(), hasher.finish())
    }
}
```

alexcrichton replied:
Agreed with @matklad on this, getting all the info into the key through the hash I think would be best here to ensure we don't forget to include anything

matklad commented on src/cargo/core/compiler/build_plan.rs:
```diff
@@ -0,0 +1,136 @@
+
+#[derive(Debug, Serialize)]
+struct Module {
+    display_name: String,
```
Instead of `display_name`, I suggest including 

```
    pub package_id: PackageId,
    pub target: Target,
```

alexcrichton replied:
It's true they won't work with `Default`, yeah, but could `Default` be removed? Using `PackageId`, `Target`, and `Profile` more closely match what Cargo is internally doing and using to represent data structures which is probably the best way to go for now to ensure it's all lossless

bors commented:
:umbrella: The latest upstream changes (presumably #5384) made this pull request unmergeable. Please resolve the merge conflicts.

mshal commented:
I've addressed most of the review feedback. The remaining issues involve the display_name hashing and storing elements of Unit in the Module. I'm a bit stumped on how to resolve it with Default, and also how to get a Unit into the update() method so that we can call get_id() and update a Module's outputs.

I also updated the BuildPlan to include a list of inputs, so an external build system will know which files cargo accessed so it can determine if the plan is out of date with respect to the input Cargo.toml files. This seemed to be fairly straightforward, but this section should also be reviewed.

After rebasing to the latest master, I noticed that cargo can no longer build mozilla-central. It fails with:

```
error: failed to resolve patches for `https://github.com/rust-lang/crates.io-index`

Caused by:
  failed to load source for a dependency on `serde_derive`

Caused by:
  Unable to update https://github.com/servo/serde?branch=deserialize_from_enums6#b8d39b01

Caused by:
  failed to update replaced source https://github.com/servo/serde?branch=deserialize_from_enums6#b8d39b01

Caused by:
  failed to parse manifest at `/home/mshal/mozilla-central-tmp/third_party/rust/serde/Cargo.toml`

Caused by:
  Feature `derive` depends on `serde_derive` which is not an optional dependency.
Consider adding `optional = true` to the dependency
```
I bisected it down to commit 4966f5394fbd562db1c7f069018a4e5781eadb7e. Ted poked around with it - it looks like it now finds the last dependency instead of the first, and serde lists serde_derive twice (once in dependencies and once in dev-dependencies), so that commit changes which one it tries to use.

mshal commented:
Is there a better way to handle the tests in a cross-platform way than what I have? Right now they are listed as [cfg(unix)] due to the paths in the JSON output and '/' not matching '\' on Windows. I also had to wildcard out some of the outputs because of the additional dSYM file in OSX, but I'd prefer if I could list the expected outputs for the platform somewhere and feed that into the expected JSON.

bors commented:
:umbrella: The latest upstream changes (presumably #5432) made this pull request unmergeable. Please resolve the merge conflicts.

alexcrichton commented:
@mshal oh the issue about optional dependencies should be resolved by https://github.com/rust-lang/cargo/pull/5454, now landed on master

alexcrichton commented on tests/testsuite/build_plan.rs:
```diff
@@ -0,0 +1,182 @@
+        .build();
+
+    assert_that(
+        p.cargo("build").arg("--build-plan"),
```
Ah one thing I've forgotten when reviewing is that we'll want the `--build-plan` argument to initially be unstable. This can be done by either renaming it to `-Z build-plan` or leaving it as this argument but also requiring `-Z unstable-options` to be passed to enable the argument

alexcrichton commented on src/cargo/core/compiler/build_plan.rs:
```diff
@@ -0,0 +1,143 @@
+use serde_json;
+
+#[derive(Debug, Default, Serialize)]
+struct Module {
```
Would a better name for this perhaps be something like `Compilation` or `Invocation`? A "module" in Rust is pretty different from what this is I believe

alexcrichton commented on src/cargo/core/compiler/build_plan.rs:
```diff
@@ -0,0 +1,143 @@
+        for arg in cmd.get_args().iter() {
+            self.args.push(
+                arg.to_str()
+                    .expect("unicode argument string required")
```
Could this use `?` and other forms of error handling to avoid a panic?

alexcrichton commented on src/cargo/core/compiler/context/mod.rs:
```diff
@@ -396,6 +426,18 @@ impl<'a, 'cfg> Context<'a, 'cfg> {
+    /// (all Cargo.toml, etc).
+    pub fn inputs(&self) -> CargoResult<Vec<PathBuf>> {
+        let mut inputs = Vec::new();
+        for id in self.packages.package_ids() {
```
I think this can run the risk of reading too much in the sense that it traverses target-specific dependencies and optional ones that aren't activated. Perhaps this could be built up as it goes along?

I also wonder, perhaps each step in the plan should depend on the manifest but not all at once? (you wouldn't want a manifest change to recompile everything, right?)

mshal replied:
Hmm, it looks like all the Cargo.toml files reported by this function are actually read by cargo during execution (according to strace). Why would we not want to list them as inputs here?

These inputs are just used to determine when we need to re-execute 'cargo --build-plan'; changing one of these inputs won't force the end-user build system to recompile everything.

I might be misunderstanding you here - perhaps we can schedule some time to chat about it.

alexcrichton replied:
Oh I see, nah if this is just rerunning the build plan that makes sense. Although I think technically the build plan can change without `Cargo.toml` being modified. For example if I move `src/lib.rs` to `src/main.rs` that implicitly removes a library from a crate and crates a binary, although it may not be worthwhile to have that much fidelity?

alexcrichton commented:
> I'm a bit stumped on how to resolve it with Default, and also how to get a Unit into the update() method so that we can call get_id() and update a Module's outputs.

I may have missed something, but it wasn't immediately apparent to me at least on first read what was causing the problems, can you expand on this a bit and maybe we can help out?

> Is there a better way to handle the tests in a cross-platform way than what I have?

Nah unfortunately it's mostly just liberally ignoring stuff like file extensions. You can use `[/]` to match the platform's path separate to handle Windows, and I think that'll work in the context of JSON matching as well

mshal commented:
>> I'm a bit stumped on how to resolve it with Default, and also how to get a Unit into the update() method so that we can call get_id() and update a Module's outputs.
> 
> I may have missed something, but it wasn't immediately apparent to me at least on first read what was causing the problems, can you expand on this a bit and maybe we can help out?

I was having problems calling get_id() on the dep_targets to build the Vec<usize> of dependencies, since the dependent Units may not have been in the BuildPlan yet via add(). However, I think this is just solved by calling plan.add() after the dependent Units are processed in compile() instead of before. Then we can know that all dependencies will already have an id in the module_map and avoid adding Default modules for dependencies that haven't been processed yet.

I haven't yet updated this to include PackageId / Target - I'm still trying to sort out how I can use those in the consumer of the JSON to produce a reasonable display. In the meantime I still have display_name, but it's not used for anything in the BuildPlan. I'll try to update this when I have time next week.

Since the module_map (now invocation_map) isn't actually needed by the JSON consumer, I split up the BuildPlan structure to have a serializable part and a part that contains the mapping. We only need the mapping to find the Invocation during update() in order to update the command with the information we get during cargo execution.

>> Is there a better way to handle the tests in a cross-platform way than what I have?
> 
> Nah unfortunately it's mostly just liberally ignoring stuff like file extensions. You can use [/] to match the platform's path separate to handle Windows, and I think that'll work in the context of JSON matching as well

Thanks! That helped me get the tests running on Windows as well.

alexcrichton commented on src/cargo/core/compiler/context/mod.rs:
```diff
@@ -63,6 +65,29 @@ pub struct Unit<'a> {
+
+        let mut hasher = SipHasher::new_with_keys(0, 0);
+        self.hash(&mut hasher);
+        format!("{}-{}", self.pkg.name(), hasher.finish())
```
I think there's a `short_hash` helper in Cargo which should assist with generating this string

alexcrichton commented:
Sounds great, thanks for continuing to work on this @mshal!

bors commented:
:umbrella: The latest upstream changes (presumably #5358) made this pull request unmergeable. Please resolve the merge conflicts.

mshal commented:
I ended up exporting the package name, package version, the unit Kind and the TargetKind so the end-user build system can display the various elements in some way to show what Invocations it is building. I initially tried to use the PackageId, but that includes the source_id URL, which is long and not very useful. With these elements we can show whether its a Host or Target invocation, and distinguish between build scripts & non-build scripts.

As a result of rebasing to master, it seems like the new home for the build_plan flag should be in BuildConfig, so I've moved it there. If I'm wrong in that assumption, let me know.

alexcrichton commented:
This all looks good to me, thanks @mshal!

@matklad did you want to take one last look as well?

matklad commented on src/cargo/core/compiler/build_plan.rs:
```diff
@@ -0,0 +1,158 @@
+            .to_str()
+            .ok_or_else(|| format_err!("unicode program string required"))?
+            .to_string()
+            .clone();
```
This `.clone` is not necessary I think: we already get an owned `String` after `to_string`. The same applies to a `clone` below as well.

matklad commented on src/cargo/core/compiler/build_plan.rs:
```diff
@@ -0,0 +1,158 @@
+                    .clone(),
+            );
+        }
+        for var in cmd.get_envs().keys() {
```
I think this could be written as `for (var, value) in cmd.get_envs()`, so as to avoid `unwrap_or_default` below.

matklad commented:
LGTM modulo couple of ultra minor nits. 

But r=me even with nits unfixed: they dont' really matter, and this awesome PR has gone through too many rebases already! 

`Invocation` struct looks great! I am rather confident making it a part of public API. The only think I am not entirely sure is that `TargetKind` is currently ambiguous, because there may be several targets with the same kind. However, lets solve this issue separately, because it is not specific to build plans: we generally don't do a very good job referring to targets. I've filed https://github.com/rust-lang/cargo/issues/5508 for that.

alexcrichton commented:
@bors: r=matklad

bors commented:
:pushpin: Commit 4016bac has been approved by `matklad`

<!-- @bors r=matklad 4016bac177024da4f99002ccef1deedc33fd759e -->

alexcrichton commented:
@bors: r=matklad

bors commented:
:pushpin: Commit a071c3a has been approved by `matklad`

<!-- @bors r=matklad a071c3ab9bf3d18f5c5b7a0c96fe670987680f9b -->

bors commented:
:hourglass: Testing commit a071c3ab9bf3d18f5c5b7a0c96fe670987680f9b with merge 13c258ecaefccaad88a173e8b8dc8517db9e8aad...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang-libs/cargo/build/1.0.4732), [status-travis](https://travis-ci.org/rust-lang/cargo/builds/377436511?utm_source=github_status&utm_medium=notification)
Approved by: matklad
Pushing 13c258ecaefccaad88a173e8b8dc8517db9e8aad to master...

jonas-schievink commented:
Does this need a tracking issue?

alexcrichton commented:
Sure yeah! @jonas-schievink want to write one up?

jonas-schievink commented:
Sure, opened https://github.com/rust-lang/cargo/issues/5579 to track this

alexcrichton commented:
Awesome, thanks!
